api/src/main/java/org/infinispan/api/reactive/Cache.java

```diff
@@ -46,90 +47,93 @@
     * @param value
     * @return Void
     */
-   CompletionStage<Void> save(K key, V value);
+   Uni<V> put(K key, V value);
+
+   /**
+    *
+    * @param key
+    * @param value
+    * @return
+    */
+   Uni<Void> set(K key, V value);
 
    /**
     * Delete the key
     *
     * @param key
-    * @return Void
+    * @return
     */
-   CompletionStage<Void> delete(K key);
+   Uni<Boolean> remove(K key);
 
    /**
     * Retrieve all keys
     *
     * @return Publisher
     */
-   Publisher<K> keys();
+   Multi<K> keys();
 
    /**
     * Retrieve all entries
     *
     * @return
     */
-   Publisher<? extends Map.Entry<K, V>> entries();
+   Multi<? extends Map.Entry<K, V>> entries();
 
    /**
     * Save many from a Publisher
     *
     * @param pairs
     * @return Void
     */
-   Publisher<WriteResult<K>> saveMany(Publisher<Map.Entry<K, V>> pairs);
+   Multi<WriteResult<K>> put(Multi<Map.Entry<K, V>> pairs);
 
    /**
     * Estimate the size of the store
     *
     * @return Long, estimated size
     */
-   CompletionStage<Long> estimateSize();
+   Uni<Long> estimateSize();
 
    /**
     * Clear the store. If a concurrent operation puts data in the store the clear might not properly work
     *
     * @return Void
     */
-   CompletionStage<Void> clear();
+   Uni<Void> clear();
 
    /**
     * Executes the query and returns a reactive streams Publisher with the results
     *
-    * @param ickleQuery query String
-    * @return Publisher reactive streams
+    * @param query query String
+    * @return
     */
-   Publisher<KeyValueEntry<K, V>> find(String ickleQuery);
+   <R> Multi<R> find(String query);
 
    /**
     * Find by QueryRequest.
     *
-    * @param queryRequest
+    * @param query
     * @return
     */
-   Publisher<KeyValueEntry<K, V>> find(QueryRequest queryRequest);
-
-   /**
-    * Executes the query and returns a reactive streams Publisher with the results
-    *
-    * @param ickleQuery query String
-    * @return Publisher reactive streams
-    */
-   Publisher<KeyValueEntry<K, V>> findContinuously(String ickleQuery);
+   <R> Query<R> query(String query);
 
    /**
     * Executes the query and returns a reactive streams Publisher with the results
     *
-    * @param queryRequest
+    * @param query query String
     * @return Publisher reactive streams
     */
-   <T> Publisher<KeyValueEntry<K, T>> findContinuously(QueryRequest queryRequest);
+   <R> Multi<R> findContinuously(String query);
 
    /**
-    * Listens to the {@link KeyValueStoreListener}
+    * Listens to the {@link KeyValueListener}
     *
     * @param listener
     * @return Publisher reactive streams
     */
-   Publisher<KeyValueEntry<K, V>> listen(KeyValueStoreListener listener);
+   Multi<KeyValueEntry<K, V>> listen(KeyValueListener listener);
```

Comment:
Why does this also return a Multi? Shouldn't we have a way to unregister the listener?

Reply:
I was thinking this would be done through `Subscription.cancel()`

Reply:
The problem is that Multi implies laziness and can be subscribed to multiple times. We only want a single listener receiving events via the listener callbacks I would expect. You could have someone subscribe to this multi multiple times receiving those values to the Multi subscription but then does the listener receive those events multiple times and what if you only cancel a Multi once?

Also I enjoyed the idea of the Handler for sync where it has a reference to the listener which makes it easier for callers to know which is which imo.

Reply:
@wburns calling `listen()` could return a lazy Multi and delay the internal listener registration to the time of the `subscribe()` call. Then cancelling the subscription would remove the listener.

Reply:
There are a few issues in this comment trail now.

1. How to cancel the listener. If we made listener removal synchronous (by not waiting for remote nodes) this would work fine by cancelling a subscription (this still has the issue in #2 though)
2. The second which I mention in my second comment is much more pressing imo. If you subscribe to the Multi multiple times those values will be duplicated in the Listener. The Listener interface has no way to delineate from multiple subscriptions.
3. Multi is lazy meaning the Listener that someone expects to get values may not receive any if they don't subscribe to it. We can chock this up to just make sure it is properly documented, but because of #2 is really weird in my books.

> @wburns calling listen() could return a lazy Multi and delay the internal listener registration to the time of the subscribe() call. Then cancelling the subscription would remove the listener.
Multi is already implied to be lazy, it can't be not and this is exactly how I would expect it to work.

Either way the main problem to me is #2 above a disconnect between the provided `KeyValueListener` which only works with a single "subscriber".

There are few ways to fix this method in my opinion.

1. `KeyValueListener` doesn't listen to change in values and this argument is used to actually configure the events passed to the Multi. (I don't like this option because we lose the typing of the events)
2. We don't return a `Multi` but rather a `ListenerHandle` to cancel it like the other methods.
3. Multi plex the Multi subscriptions and actually have a single listener registered underneath. This way multiple sources can listen to the same events and the listener is only notified once, but also means the listener wouldn't be notified unless at least one multi was subscribed (this seems really complex and unneeded).

I personally think the second option is the way this should go.

Reply:
> Either way the main problem to me is issue 3 above detailing a disconnect between the provided `KeyValueListener` which only works with a single "subscriber".

Oops, now I see what I was missing: I didn't see the `KeyValueListener` parameter, I thought `listen()` just returned a `Multi` with the events. Your answer makes a lot more sense now :)

I agree that if you already have a listener passed as a parameter, there's no point in returning a `Multi` that produces the same events (or no events?). But I'd rather remove the listener parameter and keep the `Multi`, or make them two different methods.

> 1. `KeyValueListener` doesn't listen to change in values and this argument is used to actually configure the events passed to the Multi. (I don't like this option because we lose the typing of the events)

I suggested elsewhere that the configuration of the events is not done via `KeyValueListener` methods, but via a separate metadata class. That would make the methods
`Handle listen(KeyValueListener, NotificationOptions)` and
`Multi<KeyValueEvent<K, V> listen(NotificationOptions)`

WRT event types, I agree processing all of them in a single method is not ideal. But I would say losing a bit of type safety is a fair trade for getting more expressive manipulation of events via `Multi`. I wouldn't try to extend `Multi` just to get type-safe events BTW, at least not until we have more experience with the Mutiny API.

Reply:
> `Multi<KeyValueEvent<K, V> listen(NotificationOptions)`
> 
> WRT event types, I agree processing all of them in a single method is not ideal. But I would say losing a bit of type safety is a fair trade for getting more expressive manipulation of events via `Multi`. I wouldn't try to extend `Multi` just to get type-safe events BTW, at least not until we have more experience with the Mutiny API.

Yeah, this is basically the point I got to when trying to create this API before. We are stuck configuring the listener with something disconnected from the listening of them, but that is okay. I had debated about returning an extended Multi but it becomes pretty hairy. Thinking about it more people can use things like `groupBy` to have different Flowables per event type if they aren't filtered in the options. So I guess I can be convinced to lean back this way, as it is the way I really wanted originally, but couldn't get it to the perfect spot :(

For this method I think the API should be the same for the async and non sync variant and only mutiny is different.

Reply:
> For this method I think the API should be the same for the async and non sync variant and only mutiny is different.

Since Paul mentioned that he'd want us to keep the `async=true` listener option, I wonder if we could make the sync API listeners always behave as if they had `async=true`, only running on the blocking thread pool instead of the single-threaded async-notification thread pool. An even crazier (maybe just crazy enough to work?) API for a sync listener that brings back ordering without a thread pool would be

```
AtomicBoolean done = new AtomicBoolean();
syncCache.forEachEvent((event, subscription) -> {
    if (done.get()) subscription.cancel();
    // or even
    if (done.get()) Thread.currentThread().interrupt();

    // process event
});
```

Reply:
If we want to guarantee ordering with async=true, I suspect we can just have a `LimitedExecutor` of 1 for each cache. To be honest I am not sure what the code above here is supposed to be exactly. And actually this might not even be needed per the next paragraph.

Also another side effect that we will have to watch out for is `Flowable` and `Multi` only publish entries on a single thread at a time, which may be problematic for scalability. Right now we can notify a Listener annotated method concurrently.